README.md
# Grape::Idempotency 🍇🔁

[![Gem Version](https://badge.fury.io/rb/grape-idempotency.svg)](https://badge.fury.io/rb/grape-idempotency)
[![Build Status](https://github.com/jcagarcia/grape-idempotency/actions/workflows/ruby.yml/badge.svg?branch=main)](https://github.com/jcagarcia/grape-idempotency/actions)

Gem for supporting idempotency in your [Grape](https://github.com/ruby-grape/grape) APIs.

Implementation based on the Active Internet-Draft [draft-ietf-httpapi-idempotency-key-header-04](https://datatracker.ietf.org/doc/draft-ietf-httpapi-idempotency-key-header/04/)

Topics covered in this README:

- [Installation](#installation-)
- [Basic Usage](#basic-usage-)
- [How it works](#how-it-works-)
  - [Making idempotency key header mandatory](#making-idempotency-key-header-mandatory-)
  - [Redis Storage Connectivity Issue](#redis-storage-connectivity-issue)
- [Configuration](#configuration-)
- [Changelog](#changelog)
- [Contributing](#contributing)


## Installation 🧗

Ruby 2.6 or newer is required.
[Grape](https://github.com/ruby-grape/grape) 1 or newer is required.

`grape-idempotency` is available as a gem, to install it run:

```bash
gem install grape-idempotency
```

## Basic Usage 📖

Configure the `Grape::Idempotency` class with a `Redis` instance.

```ruby
require 'redis'
require 'grape-idempotency'

redis = Redis.new(host: 'localhost', port: 6379)
Grape::Idempotency.configure do |c|
  c.storage = redis
end
```

Now you can wrap your code inside the `idempotent` method:

```ruby
require 'grape'
require 'grape-idempotency'

class API < Grape::API
    post '/payments' do
      idempotent do
        status 201
        Payment.create!({
          amount: params[:amount]
        })
      end
    end
  end
end
```

That's all! 🚀

## How it works 🤔

Once you've set up the gem and enclosed your endpoint code within the `idempotent` method, your endpoint will exhibit idempotent behavior, but this will only occur if the consumer of the endpoint includes an idempotency key in their request. (If you want to make the idempotency key header mandatory for your endpoint, check [How to make idempotency key header mandatory](#making-idempotency-key-header-mandatory-) section)

This key allows your consumer to make the same request again in case of a connection error, without the risk of creating a duplicate object or executing the update twice.

To execute an idempotent request, simply request your user to include an extra `Idempotency-Key: <key>` header as part of his request.

This gem operates by storing the initial request's status code and response body, regardless of whether the request succeeded or failed, using a specific idempotency key. Subsequent requests with the same key will consistently yield the same result, even if there were 500 errors.

Keys are automatically removed from the system if they are at least 24 hours old, and a new request is generated when a key is reused after the original has been removed. The idempotency layer compares incoming parameters to those of the original request and returns a `422 - Unprocessable Entity` status code if they don't match, preventing accidental misuse.
If a request is received while another one with the same idempotency key is still being processed the idempotency layer returns a `409 - Conflict` status

Results are only saved if an API endpoint begins its execution. If incoming parameters fail validation or if the request conflicts with another one executing concurrently, no idempotent result is stored because no API endpoint has initiated execution. In such cases, retrying these requests is safe.

Additionally, this gem automatically appends the `Original-Request` header and the `Idempotency-Key` header to your API's response, enabling you to trace back to the initial request that generated that specific response.

### Making idempotency key header mandatory ⚠️

For some endpoints, you want to enforce your consumers to provide idempotency key. So, when wrapping the code inside the `idempotent` method, you can mark it as `required`:

```ruby
require 'grape'
require 'grape-idempotency'

class API < Grape::API
    post '/payments' do
      idempotent(required: true) do
        status 201
        Payment.create!({
          amount: params[:amount]
        })
      end
    end
  end
end
```

If the Idempotency-Key request header is missing for a idempotent operation requiring this header, the gem will reply with an HTTP 400 status code with the following body:

```json
{
  "title": "Idempotency-Key is missing",
  "detail": "This operation is idempotent and it requires correct usage of Idempotency Key.",
}
```

If you want to change the error message returned in this scenario, check [How to configure idempotency key missing error message](#mandatory_header_response) section.

### Redis Storage Connectivity Issue

By default, `Redis` exceptions are not handled by the `grape-idempotency` gem.

Therefore, if an exception arises while attempting to read, write or delete data from the `Redis` storage, the gem will re-raise the identical exception to your application. Thus, you will be responsible for handling it within your own code, such as:

```ruby
require 'grape'
require 'grape-idempotency'

class API < Grape::API
    post '/payments' do
      begin
        idempotent do
          status 201
          Payment.create!({
            amount: params[:amount]
          })
        end
      rescue Redis::BaseError => e
        error!("Redis error! Idempotency is very important here and we cannot continue.", 500)
      end
    end
  end
end
```

If you want to avoid this functionality, and you want the gem handles the potential `Redis` exceptions, you have the option to configure the gem for handling these `Redis` exceptions. Please refer to the [manage_redis_exceptions](#manage_redis_exceptions) configuration property.

🚨 WARNING: If a `Redis` exception appears AFTER performing the wrapped code, nothing will be re-raised. The process will continue working and the response will be returned to the consumer of your API. However, a `409 Conflict` response can be returned to your consumer if it retried the same call with the same idempotency key. This is because the gem was not able to associate the response of the original request to the original idempotency key because those connectivity issues.

## Configuration 🪚

In addition to the storage aspect, you have the option to supply additional configuration details to tailor the gem to the specific requirements of your project.

### expires_in

As we have mentioned in the [How it works](#how-it-works-) section, keys are automatically removed from the system if they are at least 24 hours old. However, a 24-hour timeframe may not be suitable for your project. To accommodate your specific needs, you can adjust this duration by using the `expires_in` parameter for configuration:

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.expires_in = 1800
end
```

So with the cofiguration above, the keys will expire in 30 minutes.

### idempotency_key_header

As outlined in the [How it works](#how-it-works-) section, in order to perform an idempotent request, you need to instruct your users to include an additional `Idempotency-Key: <key>` header with their request. However, if this header format doesn't align with your project requirements, you have the flexibility to configure the specific header that the gem will examine to determine idempotent behavior:

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.idempotency_key_header = "x-custom-idempotency-key"
end
```

Given the previous configuration, the gem will examine the `X-Custom-Idempotency-Key: <key>` for determine the idempotent behavior.

### request_id_header

By default, this gem stores a random hex value as identifier when storing the original request and returns it in all the subsequent requests that use the same idempotency-key as `Original-Request` header in the response.

This value can be also provided by your consumer using the `X-Request-Id: <request-id>` header when performing the request to your API.

However, if you prefer to use a different format for getting the request identifier, you can configure the header to check using the `request_id_header` parameter:

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.request_id_header = "x-trace-id"
end
```

In the case above, you request your consumers to use the `X-Trace-Id: <trace-id>` header when requesting your API.

### logger, logger_level and logger_prefix

By default, the logger used by the gem is configured like `Logger.new(STDOUT)` and `INFO` level. As this gem does not log any message with `INFO` level, only `ERROR` messages will be logged.


If you want to provide your own logger, you want to change the level to `DEBUG` or you want to provide your own prefix, you can configure the gem like:

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.logger = Infrastructure::MyLogger.new
  c.logger_level = :debug
  c.logger_prefix = '[my-own-prefix]'
end
```

An example of the logged information when changing the level of the log to `DEBUG` and customizing the `logger_prefix`:

```shell
I, [2023-11-23T22:41:39.148163 #1]  DEBUG -- : [my-own-prefix] Performing endpoint "/payments" with idempotency.
I, [2023-11-23T22:41:39.148176 #1]  DEBUG -- : [my-own-prefix] Idempotency key is NOT mandatory for this endpoint.
I, [2023-11-23T22:41:39.148192 #1]  DEBUG -- : [my-own-prefix] Idempotency key received in request header "x-custom-idempotency-key" => "fd77c9d6-b7da-4966-aac8-40ee258f24aa"
I, [2023-11-23T22:41:39.148210 #1]  DEBUG -- : [my-own-prefix] Previous request information has NOT been found for the provided idempotency key.
I, [2023-11-23T22:41:39.148248 #1]  DEBUG -- : [my-own-prefix] Request stored as processing.
I, [2023-11-23T22:41:39.148261 #1]  DEBUG -- : [my-own-prefix] Performing the provided block.
I, [2023-11-23T22:41:39.148268 #1]  DEBUG -- : [my-own-prefix] Block has been performed.
I, [2023-11-23T22:41:39.148287 #1]  DEBUG -- : [my-own-prefix] Storing response.
I, [2023-11-23T22:41:39.148317 #1]  DEBUG -- : [my-own-prefix] Response stored.
I, [2023-11-23T22:41:39.148473 #1]  DEBUG -- : [my-own-prefix] Performing endpoint "/payments" with idempotency.
I, [2023-11-23T22:41:39.148486 #1]  DEBUG -- : [my-own-prefix] Idempotency key is NOT mandatory for this endpoint.
I, [2023-11-23T22:41:39.148502 #1]  DEBUG -- : [my-own-prefix] Idempotency key received in request header "x-custom-idempotency-key" => "fd77c9d6-b7da-4966-aac8-40ee258f24aa"
I, [2023-11-23T22:41:39.148523 #1]  DEBUG -- : [my-own-prefix] Request has been found for the provided idempotency key => {"path"=>"/payments", "params"=>{"locale"=>"undefined", "{\"amount\":10000}"=>nil}, "status"=>500, "original_request"=>"wadus", "response"=>"{\"error\":\"Internal Server Error\"}"}
I, [2023-11-23T22:41:39.148537 #1]  DEBUG -- : [my-own-prefix] Returning the response from the original request.
```

### manage_redis_exceptions

By default, the `grape-idempotency` gem is configured to re-raise `Redis` exceptions.

If you want to delegate the `Redis` exception management into the gem, you can configure it using the `manage_redis_exceptions` configuration property.

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.manage_redis_exceptions = true
end
```

However, this approach carries a certain level of risk. In the case that `Redis` experiences an outage, the idempotent functionality will be lost, the endpoint will behave as no idempotent, and this issue may go unnoticed.

### conflict_error_response

When providing a `Idempotency-Key: <key>` header, this gem compares incoming parameters to those of the original request (if exists) and returns a `409 - Conflict` status code if they don't match, preventing accidental misuse. The response body returned by the gem looks like:

```json
{

  "title": "Idempotency-Key is already used",
  "detail": "This operation is idempotent and it requires correct usage of Idempotency Key. Idempotency Key MUST not be reused across different payloads of this operation."
}
```

You have the option to specify the desired response body to be returned to your users when this error occurs. This allows you to align the error format with the one used in your application.

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.conflict_error_response = {
    "type": "about:blank",
    "status": 409,
    "title": "Conflict",
    "detail": "You are using the same idempotent key for two different requests"
}
end
```

In the configuration above, the error is following the [RFC-7807](https://datatracker.ietf.org/doc/html/rfc7807) format.

### processing_response

When a request with a `Idempotency-Key: <key>` header is performed while a previous one still on going with the same idempotency value, this gem returns a `409 - Conflict` status. The response body returned by the gem looks like:

```json
{

  "title": "A request is outstanding for this Idempotency-Key",
  "detail": "A request with the same idempotent key for the same operation is being processed or is outstanding."
}
```

You have the option to specify the desired response body to be returned to your users when this error occurs. This allows you to align the error format with the one used in your application.

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.processing_response = {
    "type": "about:blank",
    "status": 409,
    "title": "A request is still being processed",
    "detail": "A request with the same idempotent key is being procesed"
}
end
```

In the configuration above, the error is following the [RFC-7807](https://datatracker.ietf.org/doc/html/rfc7807) format.

### mandatory_header_response

If the Idempotency-Key request header is missing for a idempotent operation requiring this header, the gem will reply with an HTTP 400 status code with the following body:

```json
{
  "title": "Idempotency-Key is missing",
  "detail": "This operation is idempotent and it requires correct usage of Idempotency Key.",
}
```

You have the option to specify the desired response body to be returned to your users when this error occurs. This allows you to align the error format with the one used in your application.

```ruby
Grape::Idempotency.configure do |c|
  c.storage = @storage
  c.mandatory_header_response = {
    "type": "about:blank",
    "status": 400,
    "title": "Idempotency-Key is missing",
    "detail": "Please, provide a valid idempotent key in the headers for performing this operation"
}
end
```

In the configuration above, the error is following the [RFC-7807](https://datatracker.ietf.org/doc/html/rfc7807) format.

## Changelog

If you're interested in seeing the changes and bug fixes between each version of `grape-idempotency`, read the [Changelog](https://github.com/jcagarcia/grape-idempotency/blob/main/CHANGELOG.md).

## Contributing

We welcome and appreciate contributions from the open-source community. Before you get started, please take a moment to review the guidelines below.

### How to Contribute

1. Fork the repository.
2. Clone the repository to your local machine.
3. Create a new branch for your contribution.
4. Make your changes and ensure they meet project standards.
5. Commit your changes with clear messages.
6. Push your branch to your GitHub repository.
7. Open a pull request in our repository.
8. Participate in code review and address feedback.
9. Once approved, your changes will be merged.

### Development

This project is dockerized, so be sure you have docker installed in your machine.

Once you clone the repository, you can use the `Make` commands to build the project.

```shell
make build
```

You can pass the tests running:

```shell
make test
```

### Issue Tracker

Open issues on the GitHub issue tracker with clear information.

### Contributors

* Juan Carlos García - Creator - https://github.com/jcagarcia
* Carlos Cabanero - Contributor - https://github.com/Flip120

CHANGELOG.md
# Changelog
All changes to `grape-idempotency` will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.1] - (Next)

### Fix

* Your contribution here.

### Changed

* Your contribution here.

### Feature

* Your contribution here.


## [1.1.0] - 2023-12-12

### Feature

* [#20](https://github.com/jcagarcia/grape-idempotency/pull/20): Allow consumers to configure the gem for handling `Redis` exceptions - [@jcagarcia](https://github.com/jcagarcia).

## [1.0.0] - 2023-11-23

### Changed

* [#11](https://github.com/jcagarcia/grape-idempotency/pull/11): Changing error response formats - [@Flip120](https://github.com/Flip120).
* [#16](https://github.com/jcagarcia/grape-idempotency/pull/16): Changing error response code to 422 for conflict - [@Flip120](https://github.com/Flip120).

### Feature

* [#11](https://github.com/jcagarcia/grape-idempotency/pull/11): Return 409 conflict when a request is still being processed - [@Flip120](https://github.com/Flip120).
* [#15](https://github.com/jcagarcia/grape-idempotency/pull/15): Allow to mark the idempotent header as required - [@jcagarcia](https://github.com/jcagarcia).
* [#17](https://github.com/jcagarcia/grape-idempotency/pull/17): Allow to configure logger - [@jcagarcia](https://github.com/jcagarcia).

## [0.1.3] - 2023-11-07

### Fix

* [#9](https://github.com/jcagarcia/grape-idempotency/pull/9): Second calls were returning `null` when the first response was generated inside a `rescue_from`. - [@jcagarcia](https://github.com/jcagarcia).
- [#9](https://github.com/jcagarcia/grape-idempotency/pull/9): Conflict response had invalid format. - [@jcagarcia](https://github.com/jcagarcia).

## [0.1.2] - 2023-11-06

### Fix

* [#5](https://github.com/jcagarcia/grape-idempotency/pull/5): Return correct original response when the endpoint returns a hash in the body - [@jcagarcia](https://github.com/jcagarcia).

## [0.1.1] - 2023-11-06

### Fix

* [#4](https://github.com/jcagarcia/grape-idempotency/pull/4): Return `409 - Conflict` response if idempotency key is provided for same query and body parameters BUT different endpoints. - [@jcagarcia](https://github.com/jcagarcia).
* [#4](https://github.com/jcagarcia/grape-idempotency/pull/4): Use `nx: true` when storing the original request in the Redis storage for only setting the key if it does not already exist. - [@jcagarcia](https://github.com/jcagarcia).

### Changed

* [#4](https://github.com/jcagarcia/grape-idempotency/pull/4): Include `idempotency-key` in the response headers - [@jcagarcia](https://github.com/jcagarcia).
  * In the case of a concurrency error when storing the request into the redis storage (because now `nx: true`), a new idempotency key will be generated, so the consumer can check the new one seeing the headers.

## [0.1.0] - 2023-11-03

* [#1](https://github.com/jcagarcia/grape-idempotency/pull/1): Initial version - [@jcagarcia](https://github.com/jcagarcia).


spec/spec_helper.rb
require 'grape-idempotency'
require 'rack/test'
require 'mock_redis'

RSpec.configure do |config|
  config.expect_with :rspec do |c|
      c.syntax = :expect
  end

  config.order = :random

  config.include Rack::Test::Methods
end

lib/grape/idempotency/middleware/error.rb
require 'grape/middleware/base'

module Grape
  module Middleware
    class Error < Base
      def run_rescue_handler(handler, error, endpoint=nil)
        if handler.instance_of?(Symbol)
          raise NoMethodError, "undefined method '#{handler}'" unless respond_to?(handler)

          handler = public_method(handler)
        end

        if endpoint
          response = handler.arity.zero? ? endpoint.instance_exec(&handler) : endpoint.instance_exec(error, &handler)
        else
          response = handler.arity.zero? ? instance_exec(&handler) : instance_exec(error, &handler)
        end

        if response.is_a?(Rack::Response)
          update_idempotency_error_with(error, response)
          response
        else
          run_rescue_handler(:default_rescue_handler, Grape::Exceptions::InvalidResponse.new)
        end
      end

      private

      def update_idempotency_error_with(error, response)
        begin
          body = JSON.parse(response.body.join)
        rescue JSON::ParserError
          body = response.body.join
        end

        Grape::Idempotency.update_error_with_rescue_from_result(error, response.status, body)
      end
    end
  end
end

lib/grape/idempotency.rb
require 'grape'
require 'redis'
require 'logger'
require 'securerandom'
require 'grape/idempotency/version'
require 'grape/idempotency/middleware/error'

module Grape
  module Idempotency
    autoload :Helpers, 'grape/idempotency/helpers'
    Grape::Endpoint.send :include, Grape::Idempotency::Helpers

    class << self
      ORIGINAL_REQUEST_HEADER = "Original-Request".freeze

      def configure(&block)
        yield(configuration)
      end

      def restore_configuration
        clean_configuration = Configuration.new
        clean_configuration.storage = storage
        @configuration = clean_configuration
      end

      def idempotent(grape, required: false, &block)
        validate_config!
        log(:debug, "Performing endpoint \"#{grape.request.path}\" with idempotency.")
        log(:debug, "Idempotency key is #{!required ? 'NOT' : ''} mandatory for this endpoint.")

        idempotency_key = get_idempotency_key(grape.request.headers)
        log(:debug, "Idempotency key received in request header \"#{configuration.idempotency_key_header.downcase}\" => \"#{idempotency_key}\"") if idempotency_key

        grape.error!(configuration.mandatory_header_response, 400) if required && !idempotency_key
        return block.call if !idempotency_key

        stored_request = get_from_storage(idempotency_key)
        log(:debug, "Request has been found for the provided idempotency key => #{stored_request}") if stored_request
        if stored_request && (stored_request["params"] != grape.request.params || stored_request["path"] != grape.request.path)
          log(:debug, "Request has conflicts. Same params? => #{stored_request["params"] != grape.request.params}. Same path? => #{stored_request["path"] != grape.request.path}")
          log(:debug, "Returning conflict error response.")
          grape.error!(configuration.conflict_error_response, 422)
        elsif stored_request && stored_request["processing"] == true
          log(:debug, "Returning processing error response.")
          grape.error!(configuration.processing_response, 409)
        elsif stored_request
          log(:debug, "Returning the response from the original request.")
          grape.status stored_request["status"]
          grape.header(ORIGINAL_REQUEST_HEADER, stored_request["original_request"])
          grape.header(configuration.idempotency_key_header, idempotency_key)
          return stored_request["response"]
        end

        log(:debug, "Previous request information has NOT been found for the provided idempotency key.")

        original_request_id = get_request_id(grape.request.headers)
        success = store_processing_request(idempotency_key, grape.request.path, grape.request.params, original_request_id)
        if !success
          log(:error, "Request NOT stored as processing. Concurrent requests for the same idempotency key appeared.")
          grape.error!(configuration.processing_response, 409)
        else
          log(:debug, "Request stored as processing.")
        end

        log(:debug, "Performing the provided block.")

        response = catch(:error) do
          block.call
        end

        log(:debug, "Block has been performed.")

        if is_an_error?(response)
          log(:debug, "An error response was returned by the performed block. => #{response}")
          response = response[:message]
        end

        grape.header(ORIGINAL_REQUEST_HEADER, original_request_id)
        grape.body response
      rescue Redis::BaseError => e
        raise
      rescue => e
        log(:debug, "An unexpected error was raised when performing the block.")
        if !stored_request && !response
          validate_config!
          log(:debug, "Storing error response.")
          original_request_id = get_request_id(grape.request.headers)
          stored_key = store_error_request(idempotency_key, grape.request.path, grape.request.params, grape.status, original_request_id, e)
          if stored_key
            log(:debug, "Error response stored.")
            grape.header(ORIGINAL_REQUEST_HEADER, original_request_id)
            grape.header(configuration.idempotency_key_header, stored_key)
          end
        end
        log(:debug, "Re-raising the error.")
        raise
      ensure
        if !stored_request && response
          validate_config!
          log(:debug, "Storing response.")
          stored_key = store_request_response(idempotency_key, grape.request.path, grape.request.params, grape.status, original_request_id, response)
          if stored_key
            log(:debug, "Response stored.")
            grape.header(configuration.idempotency_key_header, stored_key)
          end
        end
      end

      def update_error_with_rescue_from_result(error, status, response)
        validate_config!

        stored_error = get_error_request_for(error)
        return unless stored_error

        request_with_unmanaged_error = stored_error[:request]
        idempotency_key = stored_error[:idempotency_key]
        path = request_with_unmanaged_error["path"]
        params = request_with_unmanaged_error["params"]
        original_request_id = request_with_unmanaged_error["original_request"]

        store_request_response(idempotency_key, path, params, status, original_request_id, response)
        storage.del(stored_error[:error_key])
      rescue Redis::BaseError => e
        log(:error, "Storage error => #{e.message} - #{e}")
        nil
      end

      private

      def validate_config!
        raise Configuration::Error.new("A Redis instance must be configured as cache storage") unless valid_storage?
      end

      def valid_storage?
        configuration.storage &&
        configuration.storage.respond_to?(:get) &&
        configuration.storage.respond_to?(:set) &&
        configuration.storage.respond_to?(:del)
      end

      def get_idempotency_key(headers)
        idempotency_key = nil
        headers.each do |key, value|
          idempotency_key = value if key.downcase == configuration.idempotency_key_header.downcase
        end
        idempotency_key
      end

      def get_request_id(headers)
        request_id = nil
        headers.each do |key, value|
          request_id = value if key.downcase == configuration.request_id_header.downcase
        end
        request_id || "req_#{SecureRandom.hex}"
      end

      def get_from_storage(idempotency_key)
        value = storage.get(key(idempotency_key))
        return unless value

        JSON.parse(value)
      rescue Redis::BaseError => e
        log(:error, "Storage error => #{e.message} - #{e}")
        return if configuration.manage_redis_exceptions
        raise
      end

      def store_processing_request(idempotency_key, path, params, request_id)
        body = {
          path: path,
          params: params,
          original_request: request_id,
          processing: true
        }

        storage.set(key(idempotency_key), body.to_json, ex: configuration.expires_in, nx: true)
      rescue Redis::BaseError => e
        return true if configuration.manage_redis_exceptions
        raise
      end

      def store_request_response(idempotency_key, path, params, status, request_id, response)
        body = {
          path: path,
          params: params,
          status: status,
          original_request: request_id,
          response: response
        }

        storage.set(key(idempotency_key), body.to_json, ex: configuration.expires_in, nx: false)

        idempotency_key
      rescue Redis::BaseError => e
        log(:error, "Storage error => #{e.message} - #{e}")
        idempotency_key
      end

      def store_error_request(idempotency_key, path, params, status, request_id, error)
        body = {
          path: path,
          params: params,
          status: status,
          original_request: request_id,
          error: {
            class_name: error.class.to_s,
            message: error.message
          }
        }.to_json

        storage.set(error_key(idempotency_key), body, ex: 30, nx: false)

        idempotency_key
      rescue Redis::BaseError => e
        log(:error, "Storage error => #{e.message} - #{e}")
        idempotency_key
      end

      def get_error_request_for(error)
        error_keys = storage.keys("#{error_key_prefix}*")
        return if error_keys.empty?

        error_keys.map do |key|
          request_with_error = JSON.parse(storage.get(key))
          error_class_name = request_with_error["error"]["class_name"]
          error_message = request_with_error["error"]["message"]
          
          if error_class_name == error.class.to_s && error_message == error.message
            {
              error_key: key,
              request: request_with_error,
              idempotency_key: key.gsub(error_key_prefix, '')
            }
          end
        end.first
      rescue Redis::BaseError => e
        log(:error, "Storage error => #{e.message} - #{e}")
        nil
      end

      def is_an_error?(response)
        response.is_a?(Hash) && response.has_key?(:message) && response.has_key?(:headers) && response.has_key?(:status)
      end

      def key(idempotency_key)
        "#{gem_prefix}#{idempotency_key}"
      end

      def error_key(idempotency_key)
        "#{error_key_prefix}#{idempotency_key}"
      end

      def error_key_prefix
        "#{gem_prefix}error:"
      end

      def gem_prefix
        "grape:idempotency:"
      end

      def storage
        configuration.storage
      end

      def log(level, msg)
        logger.send(level, "#{configuration.logger_prefix} #{msg}")
      end

      def logger
        return @logger if @logger

        @logger = configuration.logger
        @logger.level = configuration.logger_level
        @logger
      end

      def configuration
        @configuration ||= Configuration.new
      end
    end

    class Configuration
      attr_accessor :storage, :logger, :logger_level, :logger_prefix, :expires_in, :idempotency_key_header,
        :request_id_header, :conflict_error_response, :processing_response, :mandatory_header_response,
        :manage_redis_exceptions

      class Error < StandardError; end

      def initialize
        @storage = nil
        @logger = Logger.new(STDOUT)
        @logger_level = Logger::INFO
        @logger_prefix = "[grape-idempotency]"
        @expires_in = 216_000
        @idempotency_key_header = "idempotency-key"
        @request_id_header = "x-request-id"
        @manage_redis_exceptions = false
        @conflict_error_response = {
          "title" => "Idempotency-Key is already used",
          "detail" => "This operation is idempotent and it requires correct usage of Idempotency Key. Idempotency Key MUST not be reused across different payloads of this operation."
        }
        @processing_response = {
          "title" => "A request is outstanding for this Idempotency-Key",
          "detail" => "A request with the same idempotent key for the same operation is being processed or is outstanding."
        }
        @mandatory_header_response = {
          "title" => "Idempotency-Key is missing",
          "detail" => "This operation is idempotent and it requires correct usage of Idempotency Key."
        }
      end
    end
  end
end